src/passes/TraceCalls.cpp

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright 2024 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// Instruments the build with code to intercept selected function calls.
+// This can be e.g. used to trace allocations (malloc, free, calloc, realloc)
+// and build tools for memory usage analysis.
+// The pass supports SIMD but the multi-value feature is not supported yet.
+//
+// Instrumenting void free(void*):
+//  Before:
+//   (call $free (i32.const 64))
+//
+//  After:
+//   (local $1 i32)
+//   (block
+//     (call $free
+//       (local.tee $1
+//         (i32.const 64)
+//       )
+//     )
+//     (call $trace_free
+//       (local.get $1)
+//     )
+//   )
+//
+// Instrumenting void* malloc(int32_t) with a user-defined tracer
+// (trace_allocation):
+//  Before:
+//   (call $malloc
+//     (local.const 32))
+//
+//  After:
+//   (local $0 i32)
+//   (local $1 i32)
+//   (block (result i32)
+//     (call $trace_allocation
+//       (local.get $0)
+//       (local.tee $1
+//         (call $malloc
+//           (local.tee $0 (i32.const 2))
+//         )
+//       )
+//     )
+//   )
+//
+
+#include <set>
+
+#include "asmjs/shared-constants.h"
+#include "ir/import-utils.h"
+#include "pass.h"
+#include "support/string.h"
+#include "wasm-builder.h"
+
+namespace wasm {
+
+struct TracedFunction {
+  Name originName;
+  Name tracerName;
+  Type params;
+  Type results;
+};
+bool operator<(const TracedFunction& lhs, const TracedFunction& rhs) {
+  return lhs.originName < rhs.originName;
+}
+
+static Type strToType(const std::string& strType) {
+  if (strType.compare("i32") == 0)
```

Comment:
```suggestion
  if (strType == "i32") {
```
(`==` is more concise, and also we always require `{ }` around if bodies.